src/org/labkey/targetedms/query/QCMetricConfigurationTable.java

```diff
@@ -75,4 +91,79 @@ public static ContainerFilter getDefaultMetricContainerFilter(Container currentC
+    {
+        protected QCMetricConfigurationTableUpdateService(TableInfo queryTable, TableInfo realTable)
+        {
+            super(queryTable, realTable);
+        }
+
+        @Override
+        protected Map<String, Object> getRow(User user, Container container, Map<String, Object> keys) throws InvalidKeyException, QueryUpdateServiceException, SQLException
+        {
+            return super.getRow(user, container, keys);
+        }
+
+        @Override
+        protected Map<String, Object> insertRow(User user, Container container, Map<String, Object> row) throws DuplicateKeyException, ValidationException, QueryUpdateServiceException, SQLException
+        {
+            var insertedRow = super.insertRow(user, container, row);
+            calculateAndInsertTraceValuesForMetric((int) insertedRow.get("Id"), container, user);
+            return insertedRow;
+        }
+
+        @Override
+        public List<Map<String, Object>> insertRows(User user, Container container, List<Map<String, Object>> rows, BatchValidationException errors, @Nullable Map<Enum, Object> configParameters, Map<String, Object> extraScriptContext) throws DuplicateKeyException, QueryUpdateServiceException, SQLException
+        {
+            return super.insertRows(user, container, rows, errors, configParameters, extraScriptContext);
+        }
+
+        @Override
+        protected Map<String, Object> updateRow(User user, Container container, Map<String, Object> row, @NotNull Map<String, Object> oldRow) throws InvalidKeyException, ValidationException, QueryUpdateServiceException, SQLException
+        {
+            var updatedRow = super.updateRow(user, container, row, oldRow);
+            var metricId = (int) updatedRow.get("Id");
+            deleteTraceValueForMetric(metricId);
+            calculateAndInsertTraceValuesForMetric(metricId, container, user);
+            return updatedRow;
+        }
+
+        @Override
+        protected Map<String, Object> deleteRow(User user, Container container, Map<String, Object> oldRow) throws InvalidKeyException, QueryUpdateServiceException, SQLException
+        {
+            deleteTraceValueForMetric((Integer) oldRow.get("id"));
+            return super.deleteRow(user, container, oldRow);
+        }
+
+        @Override
+        public List<Map<String, Object>> deleteRows(User user, Container container, List<Map<String, Object>> keys, @Nullable Map<Enum, Object> configParameters, @Nullable Map<String, Object> extraScriptContext) throws InvalidKeyException, BatchValidationException, QueryUpdateServiceException, SQLException
+        {
+            return super.deleteRows(user, container, keys, configParameters, extraScriptContext);
+        }
+
+        private void deleteTraceValueForMetric(int metricId)
+        {
+            var sql = new SQLFragment("DELETE FROM " + TargetedMSManager.getTableQCTraceMetricValues() +
```

Comment:
Add a container argument and make this join through the SampleFile table to prevent an attacker that has delete access in one container being able to delete the metric values from a container where they don't have permission